File: BoardGames/Dice.cpp
#include "Dice.hpp"
#include "Utils.hpp"
#include <algorithm>


Dice::Dice() : values{1, 1, 1, 1, 1} {}

void Dice::RollAll() {
	std::transform(
		values.begin(),
		values.end(),
		values.begin(),
		[](int) {
			return rand() % FACES + 1;
		}
	);
}

void Dice::RollSelected(const std::vector<int>& indices) {
	for (int i : indices) {
		if (i > 0 && i < FACES) values[i-1] = rand() % FACES + 1;
	}
}


std::ostream& operator<<(std::ostream& os, const Dice& dice) {
	os << "[";
	for (int i : dice.values) {
		os << " " << i ;
	}
	os << " ]";
	return os;
}

int Dice::Sum() {
	int s = SumScores(values);

	return s;
}

File: BoardGamesSharp/Yahtzee/Dice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BoardGamesSharp.Yahtzee;

public class Dice
{
    private int[] values = new int[5];
    private const int Faces = 6;

    private static Random random = new Random();

    public Dice()
    {
        for (int i = 0; i < values.Length; i++)
        {
            values[i] = 1;
        }
    }


    public void RollAll()
    {
        for (int i = 0;i < values.Length; i++)
        {
            values[i] = random.Next(1, Faces + 1);
        }
    }

    public void RollSelected(List<int> indices)
    {
        foreach (int index in indices)
        {
            if (index >= 0 && index < values.Length) {
                values[index] = random.Next(1, Faces + 1);
            }
        }
    }


    public void Show()
    {
        Console.WriteLine($"[{string.Join(" ", values)}]");
    }


    public int Sum()
    {
        return values.Sum();
    }
}
